File: mono/src/main/scala/com/thunderhead/core/Reactor.scala
package com.thunderhead.core

import com.thunderhead.core.fabric.{Codec, Message, NodeHandle}
import com.thunderhead.core.impl.ManifoldInfoImpl

/**
  * Created by mike on 7/22/16.
  */
trait Reactor {
  def trampoline[T](t: T, responder: TaskListener[T])
  def trampolineError[T](err: Object, responder: TaskListener[T])
  def netSend[T](request: Message, target: NodeHandle, responder: TaskListener[T], codec: Codec[T])
  // TODO ManifoldInfoImpl -> ManifoldInfo
  def manifoldInfo(): ManifoldInfoImpl
}

File: src/main/scala/com/thunderhead/core/reactor/Reactor.java
package com.thunderhead.core.reactor;

import com.thunderhead.api.Datum;
import com.thunderhead.core.fabric.Codec;
import com.thunderhead.core.fabric.IncomingMessage;
import com.thunderhead.core.fabric.OutgoingMessage;

/**
 * Created by mike on 7/30/16.
 */
public interface Reactor {
    void run(Runnable r);

    <T> void yield(T t, TaskListener<T> continuation);

    void yieldError(TaskError err, TaskListener<?> continuation);

    // TODO: net interface class?
    <T> void netSend(Datum port, OutgoingMessage request, Codec<T> encoder);

    <T> void addNetListener(Datum port, TaskListener<IncomingMessage<T>> responder, Codec<T> decoder);

    boolean removeNetListener(Datum port);
}
